### PR TITLE
Started Framework of MatchingAlgorithm and Worked on Repeating Users

### PR DESCRIPTION
**Pull request by amanial1234**
- Created lookForMatches Function to fetch available users
- Created Compare function to compare current user to a potential match
- Created CompareSpotifyData function that compares two users and returns a double that indicates the similarity of music 
- Added an if statement to SaveSpotifyData to not generate repeating users
- Added an if statement in SaveSpotifyUserData to make sure there is a name and image from the SpotifyApi before continuing

Test Case
- Returned the Matching integer in console
- Tested for no repeating Users
Screen Shot 2022-07-20 at 10 49 55 PM

https://user-images.githubusercontent.com/103143506/180139455-1fee7766-3a3e-4d1b-8734-a92785407181.mov

### CODE REVIEW

File: MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m [hunk 1]
```diff
@@ -2,7 +2,6 @@
 #import <Parse/Parse.h>
 #import "AppDelegate.h"
 #import "MatchmakingViewController.h"
-#import <CoreData/CoreData.h>
```
Comment (amanial1234) [inline MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m:5]: This import was unnecessary

File: MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m [hunk 2]
```diff
@@ -55,7 +54,7 @@ - (void)sessionManager:(SPTSessionManager *)manager didInitiateSession:(SPTSessi
             //Gets Access Token and save user's Spotify Track Artists data
             [self saveSpotifyData:dict];
             [self saveSpotifyUserData:dict];
-    }
+        }
```
Comment (amanial1234) [inline MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m:57]: Indented. Thank you for the advice!

File: MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m [hunk 3]
```diff
@@ -96,14 +95,14 @@ -(void) getSpotifyData:(NSString *)urlString completion:(void (^)(NSDictionary *
...
-           }
-           else {
-               NSDictionary *dataDictionary = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableContainers error:nil];
-               completion(dataDictionary, nil);
-           }
-       }];
+        if (error != nil) {
+            completion(nil, error);
+        }
+        else {
```
Comment (amanial1234) [inline MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m:101]: More correction of indentation!

File: MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m [hunk 4]
```diff
@@ -150,26 +149,38 @@ - (NSDictionary *) convertSpotifyTracks:(NSArray *)tracks{
 }
 -(void) saveSpotifyData:(NSDictionary *) spotifyData{
     //Saves all Spotify Data such as there top genres,tracks,artist,album,and the artist images to the Parse Database
+    PFObject *music = [PFObject objectWithClassName:@"Music"];
```
Comment (amanial1234) [inline MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m:152]: Switched to created a Class in my database called Music instead of User

File: MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m [hunk 5]
```diff
@@ -150,26 +149,38 @@ - (NSDictionary *) convertSpotifyTracks:(NSArray *)tracks{
...
     PFUser *current = [PFUser currentUser];
-    current[@"genres"] = [NSArray arrayWithArray:spotifyData[@"genres"]];
-    current[@"tracks"] = [NSArray arrayWithArray:spotifyData[@"tracks"]];
-    current[@"artists"] = [NSArray arrayWithArray:spotifyData[@"artists"]];
-    current[@"albums"] = [NSArray arrayWithArray:spotifyData[@"albums"]];
-    current[@"images"] = [NSArray arrayWithArray:spotifyData[@"images"]];
-    [PFUser.currentUser saveInBackground];
+    PFQuery *query = [PFQuery queryWithClassName:@"Music"];
+    [query findObjectsInBackgroundWithBlock:^(NSArray *results, NSError *error) {
+        if (![[results valueForKey:@"userId"]containsObject:current.objectId]){
```
Comment (amanial1234) [inline MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m:156]: This is where I make sure that there isn't a repeating user being added to the databse

File: MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m [hunk 6]
```diff
@@ -150,26 +149,38 @@ - (NSDictionary *) convertSpotifyTracks:(NSArray *)tracks{
...
             NSString *name = userDict[@"display_name"];
             NSArray *image= userDict[@"images"];
             NSDictionary *imageDict = [image objectAtIndex:0];
-            NSString * url =[NSString stringWithFormat:imageDict[@"url"], nil];
-            [User user].profilePicture = url;
-            [User user].name = [NSString stringWithFormat: name, nil];
-                }
-        }];
+            if (!imageDict == nil){
+                [User user].name = [NSString stringWithFormat: name, nil];
```
Comment (amanial1234) [inline MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m:178]: Made sure the imageDict isn't nil before declaring objects. In later iterations, there will be a default image that will be placed if there is no image for the profile

File: MusicTaste/MusicTaste/Base.lproj/Main.storyboard [hunk 7]
```diff
@@ -282,7 +282,7 @@
                     <navigationItem key="navigationItem" id="Swl-rn-oj0"/>
                     <connections>
                         <outlet property="profileView" destination="33c-SI-ggc" id="Bq6-FU-A04"/>
-                        <outlet property="screen_name" destination="WVV-Vf-6dK" id="5qU-so-Vmm"/>
+                        <outlet property="screenName" destination="WVV-Vf-6dK" id="XW1-IC-53l"/>
```
Comment (amanial1234) [inline MusicTaste/MusicTaste/Base.lproj/Main.storyboard:285]: Renamed the variable screenName

File: MusicTaste/MusicTaste/MatchingAlgorithm.h [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface MatchingAlgorithm : NSObject
+
+@property (strong, nonatomic) NSArray *foundPosts;
++ (instancetype)shared;
++ (void) lookForMatches;
++ (void) compare:(PFObject *)potentialMatch withDictionary:(NSMutableDictionary *)matches withData:(NSManagedObject *)data;
++ (double) compareSpotifyData:(PFObject *)potentialMatch withData:(NSManagedObject *)userSpotify;
```
Comment (amanial1234) [inline MusicTaste/MusicTaste/MatchingAlgorithm.h:13]: Create the three functions used for the Matching Algorithm: look for matches, compare, and compare spotifyData

File: MusicTaste/MusicTaste/View Controllers/Login/ConnectViewController.m [hunk 9]
```diff
@@ -24,7 +22,7 @@ - (void)didTapAuthButton:(ConnectButton *)sender
     SpotifyAPIManager *api = [SpotifyAPIManager shared];
     [api setUpSpotifyWithCompletion:^(NSDictionary *data, NSError *error) {
         if (!error) {
-            //Once Authorized it the ConnectViewController will segue to the Login View Controlelr
+            //Once authorize the ConnectViewController will segue to the LoginView Controlelr
```
Comment (amanial1234) [inline MusicTaste/MusicTaste/View Controllers/Login/ConnectViewController.m:25]: Corrected grammatical error in comments

File: MusicTaste/MusicTaste/View Controllers/Login/ConnectViewController.m [hunk 10]
```diff
@@ -38,24 +36,4 @@ - (void)loadView
     [view.connectButton addTarget:self action:@selector(didTapAuthButton:) forControlEvents:UIControlEventTouchUpInside];
     self.view = view;
 }
-
-- (void)presentAlertControllerWithTitle:(NSString *)title
```
Comment (amanial1234) [inline MusicTaste/MusicTaste/View Controllers/Login/ConnectViewController.m:42]: Deleted unused function

File: MusicTaste/MusicTaste/View Controllers/MatchmakingViewController.m [hunk 11]
```diff
@@ -1,35 +1,35 @@
...
 @property (weak, nonatomic) IBOutlet UITableView *MatchTableView;
 @property (nonatomic, strong) UIRefreshControl *refreshControl;
+@property (nonatomic, strong) NSString *name;
+@property (nonatomic, strong) NSArray *profilePicture;
 
 @end
     
 @implementation MatchmakingViewController
 
-+ (instancetype)sharedMatchmaking {
```
Comment (amanial1234) [inline MusicTaste/MusicTaste/View Controllers/MatchmakingViewController.m:12]: Was not using shared instance of Matchmaking

File: MusicTaste/Podfile [hunk 12]
```diff
@@ -10,7 +10,6 @@ target 'MusicTaste' do
   pod 'Parse/UI'
   pod 'GMImagePicker'
   pod 'AFNetworking'
-  pod 'HTPressableButton'
```
Comment (amanial1234) [inline MusicTaste/Podfile:13]: Deleted unused pod

File: MusicTaste/MusicTaste/MatchingAlgorithm.m [hunk 13]
```diff
@@ -0,0 +1,93 @@
+#import "MatchingAlgorithm.h"
+#import <Foundation/Foundation.h>
+#import <Parse/Parse.h>
+#import "AppDelegate.h"
+#import "dispatch/dispatch.h"
+
+@implementation MatchingAlgorithm
+
++ (instancetype)shared {
```
Comment (amanial1234) [inline MusicTaste/MusicTaste/MatchingAlgorithm.m:9]: Creates a shared instance of Algorithm file

File: MusicTaste/MusicTaste/MatchingAlgorithm.m [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+
++ (void) lookForMatches{
+    NSMutableDictionary *matchDict = [NSMutableDictionary new];
+    PFObject *music = [PFObject objectWithClassName:@"Music"];
+    PFQuery *query = [PFQuery queryWithClassName:@"Music"];
+    PFUser *current = [PFUser currentUser];
+    //Goes through Query
+    [query findObjectsInBackgroundWithBlock:^(NSArray *spotifyData, NSError *error) {
+        for (PFObject *user in spotifyData){
+            //We make sure we are not comparing the same object
```
Comment (amanial1234) [inline MusicTaste/MusicTaste/MatchingAlgorithm.m:27]: makes sure we don't compare our own user

File: MusicTaste/MusicTaste/MatchingAlgorithm.m [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    [current saveInBackgroundWithBlock:^(BOOL succeeded, NSError * _Nullable error) {}];
+}
+
++ (double) compareSpotifyData:(PFObject *)potentialMatch withData:(NSManagedObject *)userSpotify{
+    double spotifyMatch = 0.0;
+    PFQuery *query = [PFQuery queryWithClassName:@"Music"];
+    PFUser *current = [PFUser currentUser];
+    [query whereKey:@"userId" equalTo:current.objectId];
+    //Makes sure the we are collecting information from the current user
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0), ^{
```
Comment (amanial1234) [inline MusicTaste/MusicTaste/MatchingAlgorithm.m:46]: Currently testing out dispatch asyncs to fix blocking issue

File: MusicTaste/MusicTaste/MatchingAlgorithm.m [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+                [matchDataTracks removeObjectsInArray:[potentialMatch valueForKey:@"tracks"]];
+                [matchDataAlbums removeObjectsInArray:[potentialMatch valueForKey:@"albums"]];
+                [matchDataArtists removeObjectsInArray:[potentialMatch valueForKey:@"artists"]];
+                //gets the new count of the array
+                double newGenreCount = [matchDataGenres count];
+                double newAlbumCount = [matchDataAlbums count];
+                double newTrackCount = [matchDataTracks count];
+                double newArtistCount = [matchDataArtists count];
+                //generates a integer to determine how similar the two user's music taste is - 0 for similar and 1 for not similar at all
+                spotifyMatch = (((newGenreCount/genreCount)*0.2) + ((newTrackCount/trackCount)*0.3) + ((newAlbumCount/albumCount)*0.2) + ((newArtistCount/artistCount)*0.3));
```
Comment (amanial1234) [inline MusicTaste/MusicTaste/MatchingAlgorithm.m:72]: Generate double that represents the similarity between two user's

**Review by amanial1234 — COMMENTED** [review]
I am sorry that the pull request involves comment suggestions and the current pull request for some reason it combined my previous pull request that just addressed pull request suggestions. 

In addition, I am currently working on async to make sure the query get completed fully before continuing forward. I am also working on fixing my issue on adding new users, currently its not allowing other Spotify users besides my account to be accessed. Once these two are fixed I will be able to give more accurate demonstrations of how the algorithm works.

File: MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m (hunk 3 above)
Comment (nathanielcook) [reply MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m:101]: Much easier to read when indented uniformly, thank you!

File: MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m (hunk 4 above)
Comment (nathanielcook) [reply MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m:152]: Following up on our conversation about this yesterday: Did you find any information about why this had to be done? As in, what is it about storing things in the User table/class that doesn't work exactly?

File: MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m (hunk 5 above)
Comment (nathanielcook) [reply MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m:156]: If a user changes their music preferences, would their new preferences be ignored?

File: MusicTaste/MusicTaste/AppDelegate.m [hunk 17]
```diff
@@ -47,5 +47,8 @@ - (UISceneConfiguration *)application:(UIApplication *)application configuration
 
 - (void)application:(UIApplication *)application didDiscardSceneSessions:(NSSet<UISceneSession *> *)sceneSessions {
 }
+#pragma mark - Core Data stack
+
+
```
Comment (nathanielcook) [inline MusicTaste/MusicTaste/AppDelegate.m:52]: Was this change intentional?

File: MusicTaste/MusicTaste/MatchingAlgorithm.h (hunk 8 above)
Comment (nathanielcook) [reply MusicTaste/MusicTaste/MatchingAlgorithm.h:13]: Are all three of these called publicly (i.e. called by other classes)? The `.h` file is only for public stuff that we want exposed to other classes. For anything private, that can be put in an `@interface` block in the `.m` file. This helps us hide inner complexity.

File: MusicTaste/MusicTaste/MatchingAlgorithm.m (hunk 13 above)
Comment (nathanielcook) [reply MusicTaste/MusicTaste/MatchingAlgorithm.m:9]: Can you speak to why you needed or wanted to utilize a shared instance? (You might recall that we called that a "singleton".) What would the alternative approach be?

File: MusicTaste/MusicTaste/MatchingAlgorithm.m [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+            if (![user.objectId isEqual:music.objectId]){
+                //Calls the compare function to compare each user
+                [MatchingAlgorithm compare:user withDictionary:matchDict withData:user];
+            }
+        }
+    }];
+    //sets the matches as an integer in the user's database
+    NSArray *matches = [[[matchDict keysSortedByValueUsingSelector:@selector(compare:)] reverseObjectEnumerator] allObjects];
+    music[@"matches"] = [NSArray arrayWithArray:matches];
+    [current saveInBackgroundWithBlock:^(BOOL succeeded, NSError * _Nullable error) {}];
```
Comment (nathanielcook) [inline MusicTaste/MusicTaste/MatchingAlgorithm.m:37]: I still have the concerns about "order of operations" that we talked about earlier. Were you able to verify with logging that the order of operations matches your expectations? Does `matchDict` on line 35 contain the data added to the dictionary on line 30?

File: MusicTaste/MusicTaste/MatchingAlgorithm.m (hunk 15 above)
Comment (nathanielcook) [reply MusicTaste/MusicTaste/MatchingAlgorithm.m:46]: The order of operations is really important. It looks like you're calling

lookForMatches -> compare -> compareSpotifyData

Now let's look at the order of operations

lookForMatches *findObjectsInBackgroundWithBlock* -> compare *findObjectsInBackgroundWithBlock* -> compareSpotifyData [dispatch_async] *findObjectsInBackgroundWithBlock*

So you're scheduling something to happen "in the background" while you're already running in the background. This can get messy very quickly, and can even result in deadlocks (which might be what you're referring to with "blocking issue", not sure). A deadlock is when A is waiting on B, but B is also waiting on A. So it never finishes.

Or it could be that it's not deadlocked, but it's just taking a really long time because of time complexity. In that case you are correct in that `dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0)` or similar can help, because it will do all of the work on a separate thread. Let's do some rough estimation for how the code is currently structured though. Each network call (each call to Parse or Spotify) is "expensive" because it goes over the internet. Let's estimate that each network call takes 1 second. Now let's imagine that you have 10 user profiles to examine. So for each profile you are doing a network call to Parse. Then for each of those you are doing another network call to parse for all of the other user profiles. That's O(n^2), so for 10 users that's about 100 seconds. Then you're doing a third network call in compareSpotifyData; that would be O(n^3) which would be 1000 seconds. That means it would take 16 minutes for it to complete with 10 users. Clearly we need to restructure some things.

As a start, you'll want to gather all of the data you need up front, then while you are calculating ranking scores you don't make any network calls. That will speed things up a lot, but you'll still need to address the big-O complexity.

File: MusicTaste/Podfile (hunk 12 above)
Comment (nathanielcook) [reply MusicTaste/Podfile:13]: Did you also want to remove the files in the pods directory related to that pod?

**Review by nathanielcook — COMMENTED** [review]
Pull request title and description are solid, and PR is pretty cohesive. 👍 One terminology nit about the pull request summary that I just noticed (My bad for not noticing before): Should be "Test Plan", not "Test Case". A test case is a single thing that you are testing, whereas a test plan can be made up of multiple test cases and also different types of tests. For example:

TEST PLAN

* Test that the FooViewController loads images in the list view
* Test that background color of the image matches their favorite color, and the foreground text is readable. Test case 1: Fav color is blue. Test case 2: black. Test case 3: white. Test case 4: No favorite color (null).
* Test that tapping the image navigates to BarViewController

By the way, I notice in your test plan that the match score is `0.0`; were you able to calculate non-zero match scores?

File: MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m (hunk 5 above)
Comment (amanial1234) [reply MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m:156]: If they create a new user they wouldn't but this is great to think about. I would probably implement a way to refresh their preference after a certain time.

File: MusicTaste/Podfile (hunk 12 above)
Comment (amanial1234) [reply MusicTaste/Podfile:13]: Will do!

File: MusicTaste/MusicTaste/MatchingAlgorithm.m (hunk 13 above)
Comment (amanial1234) [reply MusicTaste/MusicTaste/MatchingAlgorithm.m:9]: After further examination, I decide to still include the singleton; however, the alternative would be to have static functions and call those within other viewcontrollers since it would be public. However, I wanted the liberty to use variables within the file by calling on self and not have it static. This is why I kept the static variable and made the other functions non-static.

Comment (nathanielcook) [reply MusicTaste/MusicTaste/MatchingAlgorithm.m:9]: > the alternative would be to have static functions and call those within other viewcontrollers since it would be public

Love this! Another alternative would be to pass the instance around everywhere. (Which can sometimes get messy if it's used in many different places in the app code.)

> variables within the file by calling on self and not have it static

Out of curiosity why did you want to avoid having static class variables?
